opm/simulators/linalg/bda/CPR.cpp

```diff
@@ -0,0 +1,515 @@
+                Rmatrices.back().colIndices[col_idx] = *it;
+                col_idx++;
+            }
+        }
+    }
+
+}
+
+
+void solve_coarse_umfpack(const Matrix *A, std::vector<double> &y, std::vector<double> &x) {
+    const int N = A->N;
+    const int nnzs = A->nnzs;
+
+    using DuneMat = Dune::BCRSMatrix<Dune::FieldMatrix<double, 1, 1> >;
+    DuneMat B(N, N, nnzs, DuneMat::row_wise);
+
+    typedef DuneMat::CreateIterator Iter;
+
+    // setup sparsity pattern
+    for(Iter row = B.createbegin(); row != B.createend(); ++row){
+        int start = A->rowPointers[row.index()];
+        int end = A->rowPointers[row.index() + 1];
+        for (int idx = start; idx < end; ++idx) {
+            int col = A->colIndices[idx];
+            row.insert(col);
+        }
+    }
+
+    // set values
+    for (int row = 0; row < N; ++row) {
+        int start = A->rowPointers[row];
+        int end = A->rowPointers[row + 1];
+        for (int idx = start; idx < end; ++idx) {
+            int col = A->colIndices[idx];
+            B[row][col] = A->nnzValues[idx];
+        }
+    }
+
+    // create umfpack object
+    Dune::UMFPack<DuneMat> umfpack(B, 0);
+
+    umfpack.apply(x.data(), y.data());
+}
+
+
+template <unsigned int block_size>
+void CPR<block_size>::amg_cycle_gpu(const int level, cl::Buffer &y, cl::Buffer &x) {
+    OpenclMatrix *A = &d_Amatrices[level];
+    OpenclMatrix *P = &d_Pmatrices[level];
+    OpenclMatrix *R = &d_Rmatrices[level];
+    int Ncur = A->Nb;
+
+    if (level == num_levels - 1) {
+        // solve coarsest level
+        std::vector<double> h_y(Ncur), h_x(Ncur, 0);
+
+        events.resize(1);
+        err = queue->enqueueReadBuffer(y, CL_FALSE, 0, sizeof(double) * Ncur, h_y.data(), nullptr, &events[0]);
+        cl::WaitForEvents(events);
+        events.clear();
+        if (err != CL_SUCCESS) {
+            // enqueueWriteBuffer is C and does not throw exceptions like C++ OpenCL
+            OPM_THROW(std::logic_error, "CPR OpenCL enqueueReadBuffer error");
+        }
+
+        solve_coarse_umfpack(&Amatrices[level], h_y, h_x);
+
+        events.resize(1);
+        err = queue->enqueueWriteBuffer(x, CL_FALSE, 0, sizeof(double) * Ncur, h_x.data(), nullptr, &events[0]);
+        cl::WaitForEvents(events);
+        events.clear();
+        if (err != CL_SUCCESS) {
+            // enqueueWriteBuffer is C and does not throw exceptions like C++ OpenCL
+            OPM_THROW(std::logic_error, "CPR OpenCL enqueueWriteBuffer error");
+        }
+        return;
+    }
+    int Nnext = d_Amatrices[level+1].Nb;
+
+    cl::Buffer& t = d_t[level];
+    cl::Buffer& f = d_f[level];
+    cl::Buffer& u = d_u[level]; // u was 0-initialized earlier
+
+    // presmooth
+    double jacobi_damping = 0.72; // default value in amgcl: 0.72
+    OpenclKernels::residual(A->nnzValues, A->colIndices, A->rowPointers, x, y, t, Ncur, 1);
+    OpenclKernels::vmul(jacobi_damping, d_invDiags[level], t, x, Ncur);
```

Comment:
Maybe we should allow for a bit more flexibiility (at least a few jacobi steps)?

Anyway, interesting that one Jacobi step seems to work.

Reply:
It also seems that 0 Jacobi steps works, although it takes more linear iterations and is a little slower. I updated benchmarks in the startpost.

Reply:
We could add more runtime parameters, perhaps
`--opencl-cpr-jacobi-damping-factor` and `--opencl-cpr-jacobi-num-steps`.

Reply:
Please don't introduce any opencl-cpr specific options. We can use what is there already.

> It also seems that 0 Jacobi steps works

You mean no pre- and no post-smoothing at all in the pressure solve?

Well, I guess that is probably rather caused by the fact that ILU0 is often good enough and the course grid direction does not make it worse too much without smoothing and we only solve very inaccurately. If we would increase the the accuracy, this would probably change. According multigrid theory and my experience smoothing is needed.

Reply:
Yes, 0 Jacobi steps also converges better than only using ilu0.
I'll get a value from the `Criterion`.